src/component.ts

```diff
@@ -49,11 +63,24 @@ function makeComponent(render: RenderFunction): Creator {
     const BaseElement = (options || baseElementOrOptions as Options<P> || {}).baseElement || HTMLElement;
     const {observedAttributes = [], useShadowDOM = true, shadowRootInit = {}} = options || baseElementOrOptions as Options<P> || {};
 
+    const _observedAttributes =
+      renderer.observedAttributes || observedAttributes || [];
+
+    const props = _observedAttributes.map(item =>
+      typeof item === 'string'
+        ? {
+            type: 'legacy',
+            name: item
+          }
+        : item
+    );
+    const finalObservedAttribute = props.map((prop: any) => prop.name);
```

Comment:
You're writing two loops here that are based on the same data. This should definitely just be one loop to improve performance.